### ISSUE TITLE
Supply parameters for compression algorithms

### ISSUE THREAD

**Issue by yoavweiss** [WICG/compression#22]
Even for built in algorithms (gzip and deflate), there are various parameters that users can supply which can unlock some use cases. Examples:

* compression level - determining the tradeoff between CPU and compression ratio
* Flushing strategy - Flushing strategies can have significant impact on the tradeoff between added latency and compression ratios when compressing streams. Also, some algorithms (e.g. SSH) rely on very specific flushing strategies, so implementing them using this API may rely on being able to set those specific flushing strategies.

**Comment by ricea** [comment]
I'm am expecting to add an options bundle as the second argument to the constructor. So for example we'd have something like
```javascript
new CompressionStream('deflate', {
    level: 0.1,
    flush: 'always'
});
```
How the parameters work will be difficult to fix if we get it wrong, which is why they aren't in the initial version of the standard.

**Comment by yoavweiss** [comment]
That makes perfect sense, thanks! :)

**Comment by chris-morgan** [comment]
I imagine that will also include the use of a custom dictionary? At present for Fastmail, we use pako in a worker to compress our API request bodies, and use a custom dictionary because it makes the compression *much* more effective. I had hoped that we could look to switching it to a standard API that would compress faster with less code loaded.

Given the already-niche status of manual compression in JavaScript (for web systems specifically, I personally can’t think of having heard of even one other user, though doubtless some exist), I was a little surprised to hear of this shipping in Chrome without support for varying the compression level or providing a custom dictionary—it’s so rare that people want to do manual compression that I’d guess that a fair fraction of those that *do* use it have tuned things carefully, and so will not be able to use this new thing at this time without altering the balance.

My surprise is probably because I believe supporting at least those two parameters (level and dictionary) to be quite straightforward, with a broad approach (an options object to the constructor) being obvious, and individual options decisions that should just be made, where discussion is unlikely to affect matters. For starters, I take it as given that the available options depend wholly on the compression method selected.

`level` could reasonably be an enum, an integer or a float in the range 0–1; given JavaScript and given conventions of extant compression software, probably an integer. Its reasonable range could be 0–3 (matching FLEVEL) or 1–9 (matching most software). What the default is also varies—for FLEVEL’s 0–3, 2 is defined as the default; for compression tools with a level 1–9, some default to 8 (e.g. the zlib library) and others to 6 (e.g. gzip(1)). These numbers are, of course, fairly arbitrary anyway. You could then either leave the default unspecified, or pick 6 or 8 and run with it.

For compression, `dictionary` should probably be an String containing only ASCII, an ArrayBuffer or a Uint8Array. For decompression, you could wish to provide more than one dictionary, so perhaps `dictionary` (or `dictionaries`?) would be an object mapping Adler-32 to dictionary.

**Comment by ricea** [comment]
> I imagine that will also include the use of a custom dictionary?

Yes, that's on the roadmap, although I don't think I've specifically mentioned it here. I filed issue #27 to make it explicit.

> I was a little surprised to hear of this shipping in Chrome without support for varying the compression level or providing a custom dictionary

I believe in shipping the most uncontroversial parts of an API first. We need to assess demand to set the priority for shipping more advanced features.

> individual options decisions that should just be made, where discussion is unlikely to affect matters.

Discussion does affect matters. You gave 4 different approaches to `level` yourself. Someone else may have another suggestion which copes well with [libdeflate](https://github.com/ebiggers/libdeflate) having levels all the way up to 12, or [zopfli](https://github.com/google/zopfli) providing a higher level of compression which is extraordinarily expensive.

**Comment by noell** [comment]
The parameter should be a float in [0..1], similar to ```toDataURL```, ```toBlob```, imho.

Mapping to internal compression levels, like say libdeflate 12 for example, should be an unspecified internal implementation detail [1].

[1] ```toDataURL```, ```toBlob``` are spec'd that way.  How their parameters are mapped to the internal details of codes is not in the spec.  That was intentional -- it allows browser vendors some wiggle-room to choose what's best for their underlying implementations.

**Comment by ricea** [comment]
@noell I didn't know about the `quality` argument to `toDataURL` and `toBlob`. That's a good precedent.

I feel there should be some kind of restriction on implementations. For example, `level: 0.1` should use less CPU than `level: 0.9`, and the difference between `0.8` and `1.0` shouldn't be more than a factor of 2. The reason being that code that performs well in one browser shouldn't perform badly in another browser.

**Comment by nitedani** [comment]
The CompressionStream API's limitation of only flushing at stream closure creates significant problems for modern web applications:

1. **Server-rendered UI frameworks** require progressive streaming to enable incremental page rendering
2. **Server-Sent Events** need timely delivery of compressed events
3. **Performance metrics** like Time to First Byte and LCP are negatively impacted by buffering compressed content

Without on-demand flushing, these streaming-first use cases are forced to choose between compression and responsiveness. This limitation has led to developers implementing custom solutions with zlib (using Z_SYNC_FLUSH) or third-party libraries.

A standardized flush mechanism would align with how web applications have evolved toward streaming-centric architectures, where progressive content delivery is essential for optimal user experience.

**Comment by Mysak0CZ** [comment]
I believe the parameter's meaning, and possibly even names, should be specific to the algorithm and without any internal mapping.
After all, if one is choosing a detail such as compression level, then there already was enough research done for the specific algorithm. I believe compression steams are a rather advanced feature, so the options should be aimed at advanced users, knowledgeable about the algorithms/libraries in use.
Allowing direct pass-through of the values would also mean that any experience with fine-tuning those algorithms for other purposes can be directly used in configuring the compression stream.

My counter-example to any unification of parameters is `zstd`, where it makes sense to configure:
- Compression level, which ranges from `-7` to `22`. But levels lower than `0` are in "fast" mode, generally only for real-time applications, and levels higher than `19` require additional flag `--ultra` because of how intensive they are. So any mapping would hide those facts.
- You can configure `--long` - compression window, which at the cost of much increased memory usage can significantly improve compression. I do not believe other algorithms have a similar options.
- And of course there is dictionary option, which generally is very specific to each algorithm.

As for the API design, maybe an inspiration could be drawn from existing SubtleCrypto / Web Crypto API? That one definitely also is aimed at knowledgeable programmers. So to give an example of interface I have in mind:
```ts
interface DeflateParams {
    name: "deflate";
    level?: 0 | 1 | 2 | 3 | 4 | 5 | 6 | 7 | 8 | 9;
    strategy?: "Z_DEFAULT_STRATEGY" | "Z_FILTERED" | "Z_HUFFMAN_ONLY";
}

interface ZStdParams {
    name: "zstd";
    level?: -7 | -6 | ... | 19 | 20 | 21 | 22;
    long?: ... | 27 | 28 | 29 | 30 | 31;
    dictionary?: BufferSource;
}

type CompressionFormat = "deflate" | "deflate-raw" | "gzip" | "zstd" | ...;
type CompressionFormatWithParams = DeflateParams | ZStdParams | ...;

declare var CompressionStream: {
    new(format: CompressionFormat | CompressionFormatWithParams): CompressionStream;
};
```
By allowing both string and an object in the first argument, the extension is backwards-compatible, though not forward compatible without applications having fallback logic for when creating with object fails.
Around compatibility I still see some questions, such as whether creation with unknown option or invalid option value should fail, warn, or silently ignore it falling back to defaults/closest valid (I believe safest is failing).

There should also be consideration for allowing params for decompression stream. Mainly dictionary and for zstd also `long` can be required to decompress a stream created with them.

Finally there is a question about whether all values should be allowed, or whether to limit parameter values in the Spec itself, or let browsers transparently limit it (e.g. allowing browser to transparently lower maximum `level`/`long` if it deems the value unsuitable for the platform - e.g. trying to use `long: 31` with zstd can actually consume up to 2GB of memory, which is not an issue on PC generally, but on mobile it very well can be).

**Comment by ricea** [comment]
I prefer to add a second parameter for params rather than overload the first parameter. So
```js
const compressor = new CompressionStream("zstd", { level: 22 });
```
rather than
```js
const compressor = new CompressionStream({ name: "zstd", level: 22 });
```
I feel this makes adding options to an existing call more intuitive.

I am leaning towards using the de-facto standard zlib-style naming for compression levels, for the reason that both Brotli and Zstd also use it, and so it is most likely to be familiar to people.

Being able to use alternate implementations is important. Chrome is trying to move more code to memory-safe languages, so if there were suitable Rust implementations Chrome might switch to them. This doesn't mean we can't use zlib-style naming for levels, just that I want to permit flexibility in how implementations interpret them.

By the way, all of gzip, Brotli and Zstd can use a stream without any specific format as a dictionary. Brotli has experimental support for a "serialized" shared dictionary, which we might possibly support in future with a Brotli-specific `dictionaryType` attribute.
